x/mongo/driver/topology/server.go

```diff
@@ -437,119 +526,157 @@ func (s *Server) updateDescription(desc description.Server) {
+			// The connection is already in a streaming state, so we stream the next response.
+			err = baseOperation.StreamResponse(s.heartbeatCtx, heartbeatConn)
+		case previousDescription.TopologyVersion != nil:
+			// The server supports the streamable protocol. Set the socket timeout to
+			// connectTimeoutMS+heartbeatFrequencyMS and execute an awaitable isMaster request. Set conn.canStream so
+			// the wire message will advertise streaming support to the server.
+
+			// Calculation for maxAwaitTimeMS is taken from time.Duration.Milliseconds (added in Go 1.13).
+			maxAwaitTimeMS := int64(s.cfg.heartbeatInterval) / 1e6
+			s.conn.setSocketTimeout(s.cfg.heartbeatTimeout + s.cfg.heartbeatInterval)
+			baseOperation = baseOperation.TopologyVersion(previousDescription.TopologyVersion).
+				MaxAwaitTimeMS(maxAwaitTimeMS)
+			s.conn.setCanStream(true)
+			err = baseOperation.Execute(s.heartbeatCtx)
+		default:
+			// The server doesn't support the awaitable protocol. Set the socket timeout to connectTimeoutMS and
+			// execute a regular heartbeat without any additional parameters.
 
-		// we do a retry if the server is connected, if succeed return new server desc (see below)
-		if err != nil {
-			saved = err
-			conn = nil
-			if wrappedConnErr := unwrapConnectionError(err); wrappedConnErr != nil {
-				s.pool.clear()
-				// If the server is not connected, give up and exit loop
-				if s.Description().Kind == description.Unknown {
-					break
-				}
+			s.conn.setSocketTimeout(s.cfg.heartbeatTimeout)
+			err = baseOperation.Execute(s.heartbeatCtx)
+		}
+		if err == nil {
+			tempDesc := baseOperation.Result(s.address)
+			descPtr = &tempDesc
+		} else {
+			// Close the connection here rather than below so we ensure we're only closing a connection if one was
+			// successfully created.
+			if s.conn != nil {
+				_ = s.conn.close()
 			}
-			continue
 		}
+	}
 
-		desc = *descPtr
-		delay := time.Since(now)
-		desc = desc.SetAverageRTT(s.updateAverageRTT(delay))
-		desc.HeartbeatInterval = s.cfg.heartbeatInterval
-		set = true
+	if descPtr != nil {
+		// The check was successful. Set the average RTT and return.
+		desc := *descPtr
+		desc.AverageRTT = s.rttMonitor.getRTT()
 
-		break
+		return desc, nil
 	}
 
-	if !set {
-		desc = description.NewServerFromError(s.address, saved)
+	if s.checkWasCancelled() {
+		// If the previous check was cancelled, we don't want to clear the pool. Return a sentinel error so the caller
+		// will know that an actual error didn't occur.
+		return emptyDescription, errCheckCancelled
 	}
 
-	return desc, conn
-}
-
-func (s *Server) updateAverageRTT(delay time.Duration) time.Duration {
-	if !s.averageRTTSet {
-		s.averageRTT = delay
-	} else {
-		alpha := 0.2
-		s.averageRTT = time.Duration(alpha*float64(delay) + (1-alpha)*float64(s.averageRTT))
-	}
-	return s.averageRTT
+	// An error occurred. We clear the pool for all errors and return an Unknown description.
+	s.rttMonitor.reset()
+	s.pool.clear()
```

Comment:
This is a bug I need to address in a new spec ticket. We MUST clear the pool only _after_ marking the server Unknown. Clearing the pool before updating the Topology causes the race described here: https://jira.mongodb.org/browse/PYTHON-2236?focusedCommentId=3076780&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-3076780